### ISSUE TITLE
Optimize Generic new() Constraint

### ISSUE THREAD

**Issue by RehanSaeed** [dotnet/roslyn#22464]
[Jon Skeet](https://codeblog.jonskeet.uk/2011/08/22/optimization-and-generics-part-1-the-new-constraint/) has a whole blog post describing how the generic `new()` constraint calls `Activator.CreateInstance` which uses reflection, so it ends up being slow.

Workarounds for this at the moment are to use `default(T)` for structs or creating a `Func<T>` lambda using expression trees like so:

```
Expression.Lambda<Func<T>>(Expression.New(typeof(T))).Compile();
```

Ideally `new()` can be optimized so that these workarounds are not necessary.

**Comment by sharwell** [comment]
:link: Related to (and possible duplicate of) dotnet/coreclr#7592.

**Comment by 4creators** [comment]
> Related to (and possible duplicate of) dotnet/coreclr#7592.

@sharwell  imho it seems to be different problem. Here the problem is about using `Activator.CreateInstance` by `new() generic constraint` which may be solved by avoiding use of `Activator.CreateInstance` instead of optimizing `Activator.CreateInstance` what seems to be non trivial according to https://github.com/dotnet/coreclr/issues/7592.

**Comment by AlexeiScherbakov** [comment]
Also, new() require **public** costructor, but sometimes it needed to have **internal** constructor
workaround is:
`internal static class ObjectConstructorsCache<T>
	{
		private static readonly Func<T> _ctor;

		static ObjectConstructorsCache()
		{
			var type = typeof(T);
			ConstructorInfo ctor = type.GetConstructor(BindingFlags.Public | BindingFlags.Instance, null, Type.EmptyTypes, null);
			if (null == ctor)
			{
				ctor = type.GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, null, Type.EmptyTypes, null);
			}
			LambdaExpression lambda = Expression.Lambda(typeof(Func<T>), Expression.New(ctor));
			var del = lambda.Compile();
			_ctor = (Func<T>) del;
		}

		public static Func<T> Ctor
		{
			get { return _ctor; }
		}
	}`

**Comment by jaredpar** [comment]
> Ideally new() can be optimized so that these workarounds are not necessary.

Can you elaborate on how exactly you think the compiler can improve it's code generation here. Specifically what examples can be optimized and what approach you think we could take? 

The general case of an unconstrained type parameter can't be reasonably optimized by the compiler. The runtime is in a much better position to optimize such calls. Possibly we should be using a different API than `Activator.CreateInstance` but that API would need to be provided by the runtime.

The more specific case of a constrained type parameter has issues. Consider the case of `where T : struct`. Even optimizing that by just changing `new T()` to `default(T)` causes compatibility issues. Most versions of `Activator.CreateInstance` will actually call a `struct` constructor if it's defined in IL. Moving to `default(T)` would break anyone that did that. Even though it's not legal in C# to define a parameterless constructor on a `struct` it's just fine in IL and will be respected in many cases ([Jon Skeet](https://codeblog.jonskeet.uk/2008/12/10/value-types-and-parameterless-constructors/). 

In short: there really isn't a good way for the compiler to optimize this and even if we chose to we'd likely break behavior.
